### PR TITLE
[master] Fix Flaky LazyLoaderRefreshFileMappingTest

### LINKED ISSUE #64567: [TEST FAILURE] Flaky `LazyLoaderRefreshFileMappingTest` Test
Please paste the link from the specific build where the failure first started: **This occurred in CI for a private fork of the `salt` repository, but is something that we felt would be useful upstream.**

The error seen was related to a mismatch of an [assertion](https://github.com/saltstack/salt/blob/0892d076846525e694f8aedb6217884ca75845f3/tests/unit/test_loader.py#L1724) between the number of calls to a function and the number of times a lock was acquired.

In our local CI environment, the `threading.RLock.__enter__` function is called more often than the `_refresh_file_mapping` function, which leads to occasional failures (the full text of the issue is below).

```bash
tests/unit/test_loader.py:1722 (LazyLoaderRefreshFileMappingTest.test_lazyloader_refresh_file_mapping_called_with_lock_at___init__)
2 != 11

self = <tests.unit.test_loader.LazyLoaderRefreshFileMappingTest testMethod=test_lazyloader_refresh_file_mapping_called_with_lock_at___init__>

    def test_lazyloader_refresh_file_mapping_called_with_lock_at___init__(self):
        func_mock = MagicMock()
        lock_mock = MagicMock()
        lock_mock.__enter__ = MagicMock()
        self.LOADER_CLASS._refresh_file_mapping = func_mock
        with patch("threading.RLock", MagicMock(return_value=lock_mock)):
            loader = self.__init_loader()
        lock_mock.__enter__.assert_called()
        func_mock.assert_called()
>       assert len(func_mock.call_args_list) == len(lock_mock.__enter__.call_args_list)
E       AssertionError: assert 2 == 11
E        +  where 2 = len([call(), call()])
E        +    where [call(), call()] = <MagicMock id='140237436141920'>.call_args_list
E        +  and   11 = len([call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call()])
E        +    where [call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call(),\n call()] = <MagicMock name='mock.__enter__' id='140237435083984'>.call_args_list
E        +      where <MagicMock name='mock.__enter__' id='140237435083984'> = <MagicMock id='140237435272736'>.__enter__

tests/unit/test_loader.py:1732: AssertionError
```

### PR DESCRIPTION
**Pull request by cianyleow**
### What does this PR do?
Move creation of the lock to a private method to isolate mocking behaviour from main threading library.

### What issues does this PR fix or reference?
Fixes: [64567](https://github.com/saltstack/salt/issues/64567)

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
- [N/A] Docs
- [YES] Changelog - `64567.fixed.md`
- [YES] Tests written/updated

### Commits signed with GPG?
**Yes**/No

### CODE REVIEW

**Review by Ch3LL — DISMISSED** [review]
This looks good but I certainly want to get @dwoz and @s0undt3ch 's review here

File: tests/unit/test_loader.py
Comment (s0undt3ch) [inline tests/unit/test_loader.py:1721]:
```suggestion
        with patch.object(self.LOADER_CLASS, "_get_lock", MagicMock(return_value=lock_mock)):
            loader = self.__init_loader()
```

Comment (s0undt3ch) [reply tests/unit/test_loader.py:1721]: This is untested code, but we need to use `patch` to revert the changes after the test completes

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by s0undt3ch — APPROVED** [review]

**Comment by cianyleow** [comment]
@s0undt3ch thanks for the speedy reviews! I think I fixed the CI issues I noticed the other day, but will need a re-run to confirm

**Review by s0undt3ch — APPROVED** [review]

**Comment by cianyleow** [comment]
@s0undt3ch I think the failing CI checks are unrelated (I see the same Windows tests are failing on [`master`](https://github.com/saltstack/salt/actions/runs/5447608111/jobs/9911153644) for the same reason). Is there anything I can do to help fix them to get this merged?

**Comment by cianyleow** [comment]
Sorry to keep bugging @s0undt3ch, could I get a re-run of my CI? I think the previous Windows issue is resolved (and I've rebased), so it pass this time!

**Review by s0undt3ch — APPROVED** [review]

**Comment by cianyleow** [comment]
Hi again @s0undt3ch, I realised this never got merged and now its out of date again, so I've rebased on the latest `master` and force-pushed to clear up the history. Would you be able to run CI again on it please?

**Review by Ch3LL — APPROVED** [review]

**Comment by cianyleow** [comment]
Hey @Ch3LL not sure on the best way to get this merged as I've been in this cycle of either having failing tests or an out of date branch - would you like me to rebase on the latest `master` so that it can be re-run? The failing tests are related to `Windows 2016` which look similar to other ones I've had flake on me in the past.

**Comment by Ch3LL** [comment]
Apologies, I have the 3007.0 label on this so it's on my radar. We are currently waiting on a merge forward for fixing all the tests. I'll make sure to follow up here to update the branch once that is merged into master.

**Pull request merged**
